src/core/ConfigurableCorePool.ts

```diff
@@ -0,0 +1,126 @@
+import BN from "bn.js";
+import { TickManager } from "../manager/TickManager";
+import { PositionManager } from "../manager/PositionManager";
+import { PoolState } from "../model/PoolState";
+import { Record } from "../entity/Record";
+import { PoolConfig } from "../entity/PoolConfig";
+import { CorePool } from "./CorePool";
+import { ActionType } from "../enum/ActionType";
+import { Serializer } from "../util/Serializer";
+import { Snapshot } from "../entity/Snapshot";
+
+export class ConfigurableCorePool extends CorePool {
+  private poolState: PoolState;
+  private postProcessorCallback: (
+    this: ConfigurableCorePool,
+    recordId: string,
+    actionType: ActionType,
+    actionParams: object
+  ) => void = function (
+    this: ConfigurableCorePool,
+    recordId: string,
+    actionType: ActionType,
+    actionParams: object
+  ) {};
+
+  constructor(poolConfig: PoolConfig);
+  constructor(poolConfig: PoolConfig, poolState: PoolState);
+  constructor(
+    poolConfig: PoolConfig,
+    poolState: PoolState = new PoolState(null, poolConfig)
+  ) {
+    super(
+      poolConfig.token0,
+      poolConfig.token1,
+      poolConfig.fee,
+      poolConfig.tickSpacing
+    );
+    if (poolState.hasBaseSnapshot()) {
+      let poolConfig = poolState.poolConfig;
+      let state = <Snapshot>poolState.baseSnapshot;
+      super(
+        poolConfig.token0,
+        poolConfig.token1,
+        poolConfig.fee,
+        poolConfig.tickSpacing,
+        state.token0Balance,
+        state.token1Balance,
+        state.sqrtPriceX96,
+        state.tickCurrent,
+        state.feeGrowthGlobal0X128,
+        state.feeGrowthGlobal1X128,
+        <TickManager>Serializer.deserialize(TickManager, state.ticks_json),
+        <PositionManager>(
+          Serializer.deserialize(PositionManager, state.positions_json)
+        )
+      );
+    }
+    this.poolState = poolState;
+  }
+
+  mint(tickLower: number, tickUpper: number, amount: BN) {
+    // TODO
+    this.postProcess(ActionType.MINT, {
+      tickLower: tickLower,
+      tickUpper: tickUpper,
+      amount: amount,
+    });
+  }
+
+  burn(tickLower: number, tickUpper: number, amount: BN) {
+    // TODO
+  }
+
+  collect(
+    tickLower: number,
+    tickUpper: number,
+    amount0Requested: BN,
+    amount1Requested: BN
+  ) {
+    // TODO
+  }
+
+  swap(zeroForOne: boolean, amountSpecified: BN, sqrtPriceLimitX96: BN) {
+    // TODO
+  }
+
+  updatePostProcessor(
+    callback: (
+      this: ConfigurableCorePool,
+      recordId: string,
+      actionType: ActionType,
+      actionParams: object
+    ) => void
+  ) {
+    this.postProcessorCallback = callback;
+  }
+
+  takeSnapshot(): string {
+    // TODO
+    return "0";
+  }
+
+  recover(snapshotId: string) {
+    // TODO
+  }
+
+  private doRecord(actionType: ActionType, actionParams: object): Record {
```

Comment:
Consider renaming the function to `applyRecord`

Reply:
Just to make sure, does `apply` here means to write down(the record)?

Reply:
Nope, and me misunderstanding the meaning of the function is the problem caused by the naming.
In this case, `PersistRecord` might be more appropriate.

Reply:
Let me rephrase, I kind of think `apply` is just fine because this function is to build a record then append it to the current `PoolState`. Do we share the same idea?

Reply:
Oh then my original understanding was correct, `applyRecord` means applying the delta on top of a state.